seisidd/experiment.py

```diff
@@ -401,16 +544,845 @@ def scan_closure():
+            epics.caput("PV_DET:cam1:FrameType_RBV.ONST", "/exchange/data")
+            epics.caput("PV_DET:cam1:FrameType_RBV.TWST", "/exchange/data_white_post")
+            epics.caput("PV_DET:cam1:FrameType_RBV.THST", "/exchange/data_dark")
+            # set the layout file for cam
+            # TODO:  need to udpate with acutal config files for 6-ID-D
+            _current_fp = str(pathlib.Path(__file__).parent.absolute())
+            _attrib_fp = os.path.join(_current_fp, 'config/PG2_attributes.xml')
+            _layout_fp = os.path.join(_current_fp, 'config/tomo6bma_layout.xml')
+            det.cam.nd_attributes_file.put(_attrib_fp)
+            det.hdf1.xml_file_name.put(_layout_fp)
+            # turn off the problematic auto setting in cam
+            det.cam.auto_exposure_auto_mode.put(0)  
+            det.cam.sharpness_auto_mode.put(0)
+            det.cam.gain_auto_mode.put(0)
+            det.cam.frame_rate_auto_mode.put(0)
+        else:
+            raise ValueError(f"Invalide mode, {mode}")
+        return det
+
+    ############################
+    ## Near Field Calibration ##
+    ############################
+    #   NOT to be used in scan plans
+    def calibration(self):
+        """Image calibration for the two NF z positions"""
+        det     = self.nf_det
+        # add the z motor?
+        # add the beamstp motor?
+        pass
+
+    # ----- pre-defined scan plans starts from here
+    @bpp.run_decorator()
+    def fly_scan(self, cfg_nf):
+        """
+        Collect projections with fly motion
+        """
+        det     = self.nf_det
+        psofly  = self.fly_control
+        
+        # TODO:
+        #   Need to set up FS control for the scan
+        #   During fly scan, the FS is always open
+
+        # TODO:
+        #   The fields need to be updated for 6-ID-D
+        yield from bps.mv(det.hdf1.nd_array_port, 'PG1')
+        yield from bps.mv(det.tiff1.nd_array_port, 'PG1')
+    
+        # we are assuming that the global psofly is available
+        yield from bps.mv(
+            psofly.start,           cfg_nf['omega_start'],
+            psofly.end,             cfg_nf['omega_end'],
+            psofly.scan_delta,      abs(cfg_nf['omega_step']),
+            psofly.slew_speed,      cfg_nf['slew_speed'],
+        )
+        # taxi
+        yield from bps.mv(psofly.taxi, "Taxi")
+        yield from bps.mv(
+            det.cam.num_images, cfg_nf['n_projections'],
+            det.cam.trigger_mode, "Overlapped",
+        )
+        # start the fly scan
+        yield from bps.trigger(det, group='fly')
+        yield from bps.abs_set(psofly.fly, "Fly", group='fly')
+        yield from bps.wait(group='fly')
+
+    @bpp.stage_decorator([det])         
+    @bpp.run_decorator()
+    def scan_singlelayer(self, cfg_nf, _layer_number):
+        # TODO:
+        #   Somewhere we need to check the light status, or, add a suspender?
```

Comment:
yes, but we cannot instantiate anything until the system is in-place (devices and soft-IOCs).

Reply:
Sure, this is a reminder for myself